### PR TITLE
CPU plugin selective build

### PR DESCRIPTION
**Pull request by vladislav-volkov**
Conditional compilation applied to the CPU plugin.

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp [hunk 1]
```diff
@@ -71,38 +71,42 @@ void cpu_convert(const void *srcPtr, void *dstPtr, Precision srcPrc, Precision d
...
+    MKLDNN_CNV(I64, FP32), MKLDNN_CNV(I64, BF16),  MKLDNN_CNV(I64, BOOL),
+    MKLDNN_CNV(FP32, U8),  MKLDNN_CNV(FP32, I8),   MKLDNN_CNV(FP32, U16),
+    MKLDNN_CNV(FP32, I16), MKLDNN_CNV(FP32, I32),  MKLDNN_CNV(FP32, U64),
+    MKLDNN_CNV(FP32, I64), MKLDNN_CNV(FP32, BF16), MKLDNN_CNV(FP32, BOOL),
+    MKLDNN_CNV(BF16, U8),  MKLDNN_CNV(BF16, I8),   MKLDNN_CNV(BF16, U16),
+    MKLDNN_CNV(BF16, I16), MKLDNN_CNV(BF16, I32),  MKLDNN_CNV(BF16, U64),
+    MKLDNN_CNV(BF16, I64), MKLDNN_CNV(BF16, FP32), MKLDNN_CNV(BF16, BOOL),
+    MKLDNN_CNV(BOOL, U8),  MKLDNN_CNV(BOOL, I8),   MKLDNN_CNV(BOOL, U16),
+    MKLDNN_CNV(BOOL, I16), MKLDNN_CNV(BOOL, I32),  MKLDNN_CNV(BOOL, U64),
+    MKLDNN_CNV(BOOL, I64), MKLDNN_CNV(BOOL, FP32), MKLDNN_CNV(BOOL, BF16));
```
Comment (ilya-lavrenov) [inline inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp:106]: looks like here we have significantly more instantiations, is not it?

**Review by ilya-lavrenov — APPROVED** [review]

File: inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp [hunk 2]
```diff
@@ -71,38 +71,42 @@ void cpu_convert(const void *srcPtr, void *dstPtr, Precision srcPrc, Precision d
...
-            break;
-        case Precision::BOOL:
-            convertFrom<PrecisionTrait<Precision::BOOL>::value_type>(srcPtr, dstPtr, dstPrc, size);
-            break;
-        default:
-            THROW_IE_EXCEPTION << "cpu_convert can't convert from: " << srcPrc << " precision";
-    }
+    ConvertContext ctx = { srcPtr, dstPtr, size, false };
+
+    OV_SWITCH(MKLDNNPlugin, ConvertPrecision, ctx, std::tie(srcPrc, dstPrc),
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp:76]: Yes we can, but unfortunately this leads to ineffective disabling of unused code.
For example, given two switches.

switch(k1) {
case 1: ...
case 2: ...
}

switch(k2) {
case 1: ...
case 2: ...
}
Suppose we only have to enable cases (1,1) and (2,2).
In the first switch, we enable 1 and 2, and in the second switch, we also enable cases 1 and 2. Unfortunately this leads to enabling of undesired cases (1,2) and (2,1).

File: inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp [hunk 3]
```diff
@@ -24,45 +28,41 @@ void convert(const void *srcPtr, void *dstPtr, const size_t size) {
...
+    void operator()(ConvertContext & ctx) {
+        convert<src_t, dst_t>(ctx.srcPtr, ctx.dstPtr, ctx.size);
+        ctx.converted = true;
     }
-}
+};
+
+}   // namespace
+
+#define MKLDNN_CNV(ST, DT) OV_CASE2(Precision::ST, Precision::DT, PrecisionInfo<Precision::ST>::value_type, PrecisionInfo<Precision::DT>::value_type)
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/common/cpu_convert.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/mkldnn_selective_build.h [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// Copyright (C) 2018-2020 Intel Corporation
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/mkldnn_selective_build.h]: fixed

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**